addons/stock/models/product.py

```diff
@@ -370,6 +371,13 @@ def action_view_stock_move_lines(self):
         action['domain'] = [('product_id', '=', self.id)]
         return action
 
+    def action_view_related_putaway_rules(self):
+        self.ensure_one()
+        domain = ['|',
+            ('product_id', '=', self.id),
+            ('category_id', '=', self.product_tmpl_id.categ_id.id)]
```

Comment:
``` python
domain = [
    '|',
        ('product_id', '=', self.id),
        ('category_id', '=', self.product_tmpl_id.categ_id.id),
]
```
is more readable
also i would create an helper to return the action instead of calling the public method of product_tmpl with an argument

in the product_template you could 
```
@api.model
def _get_action_view_related_putaway_rules(domain):
return {
            'name': _('Putaway Rules'),
            'type': 'ir.actions.act_window',
            'res_model': 'stock.putaway.rule',
            'view_type': 'list',
            'view_mode': 'list',
            'domain': domain,
        }
```
and call this helper in the two `action_view_related_putaway_rules`